qemu: add override qemu_full

bb2020 opened the pull request:
###### Motivation for this change
#41615

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented:
Please remove the master merge commit.

andir commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -19,7 +19,7 @@
 , openGLSupport ? sdlSupport, mesa, epoxy, libdrm
 , virglSupport ? openGLSupport, virglrenderer
-, smbdSupport ? false, samba
+, smbdSupport ? true, samba
```
This should not be unconditional as it pull in samba as a dependency. Set it to `!nixosTestRunner` instead if you must.

Instead of doing this I'd rather have a `qemu_full` package that enables all the optional features. Things like `cephSupport` are probably more or just as likely to be used as samba.

bb2020 replied:
Good idea. I did not realize how big `samba` package is. Is it going to be based on `qemu` or `qemu_kvm` ?

SuperSandro2000 replied:
I think they are the same package and if not: I would use qemu_kvm because full.

bb2020 replied:
`qemu` is the fuller one, so I used that and updated the pull request.

andir reviewed (changes requested):
We shouldn't set it to true unconditionally. I left a comment in the code review.

Also please add the motivation you detailed in your PR description to the commit changing the knob. It will make it much easier for someone (likely neither me nor you) to figure out the *why* when debugging/refactoring something.

bb2020 committed:
pkgs/qemu: add qemu_full package

qemu_full enables samba and ceph support because otherwise enabling
them triggers recompilation of qemu package that takes a long time.
Similar options can be enabled later on. qemu_full is based on complete
qemu package, so hostCpuOnly is not enabled.

andir commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18870,6 +18870,7 @@ in
   prototool = callPackage ../development/tools/prototool { };
 
   qemu_kvm = lowPrio (qemu.override { hostCpuOnly = true; });
+  qemu_full = lowPrio (qemu.override { smbdSupport = true; cephSupport = true; });
```
I fear that we might have a feature drift by having to manage feature flags both here and in the QEMU derivation... Any ideas?

bb2020 replied:
Isn't this how it is supposed to be? I've seen similar other packages like `wine`.

andir reviewed (approved):
I am still not super happy about the amounts of QEMU packages that we have in nixpkgs but that isn't your fault nor something that I can ask you to change.

bb2020 commented:
Thanks for merging despite the confusion!
